### PR TITLE
`StoreKit1Wrapper`: process transactions in a background thread

### LINKED ISSUE #2107: Processing under the main thread hangs the UI
- [X] I have updated Purchases SDK to the latest version
- [X] I have read the [Contribution Guidelines](https://github.com/RevenueCat/purchases-ios/blob/main/Development/CONTRIBUTING.md)
- [X] I have searched the [Community](https://community.revenuecat.com)
- [X] I have read [docs.revenuecat.com](https://docs.revenuecat.com/)
- [X] I have searched for [existing Github issues](https://github.com/RevenueCat/purchases-ios/issues)

**Describe the bug**
I'm investigating another bug, but trying some recommendations from yours, like using sk1 and enabling log level to debug, I find that the application crashes by memory or hangs the UI.

1. Environment
   1. Platform: iOS
   2. SDK version: 4.14.3
   3. StoreKit 2 (_disabled with `useStoreKit2IfEnabled(false)`_) (Y/N): Which SDK uses by default (I think is SK1 right now)
   4. OS version:  x.x.x
   5. Xcode version: 14.1.0
   6. How widespread is the issue. Percentage of devices affected. We don't know because depends on transactions.
2. Debug logs:
[logsRevenuecat.zip](https://github.com/RevenueCat/purchases-ios/files/10120957/logsRevenuecat.zip)

3. Steps to reproduce:

The bug is reproduced when the SDK starts to process transactions. I think that part of the process is done inside of the Main actor, using the main thread for that. When It processes only a few transactions, all is working as expected, but when the number of transactions is high, the UI hangs, and memory increases exponentially.

4. **Other information** (e.g. stacktraces, related issues, suggestions how to fix, links for us to have context, eg. stackoverflow, etc.)

image

I don't know if it makes sense, but maybe there's a way to use a background thread to process them or maybe create a global actor that uses other queues (custom if it's possible). Also, processing a great number of transactions could take time, maybe there's a way to process a prefix or suffix of 50, ordering them by date && success result and don't communicate the rest of them directly to back doing batches.

**Additional context**

### PR DESCRIPTION
**Pull request by NachoSoto**
Fixes #2107 and [SDKONCALL-180]. That's a rare issue where a user ends up with thousands of transactions.
That exposed the fact that we're processing all of these transactions on the main thread. This fixes that.

[SDKONCALL-180]: https://revenuecats.atlassian.net/browse/SDKONCALL-180?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: Sources/Purchasing/StoreKit1/StoreKit1Wrapper.swift [hunk 1]
```diff
@@ -61,17 +61,19 @@ class StoreKit1Wrapper: NSObject {
     }
 
     private let paymentQueue: SKPaymentQueue
+    private let operationDispatcher: OperationDispatcher
 
-    init(paymentQueue: SKPaymentQueue) {
+    init(paymentQueue: SKPaymentQueue, operationDispatcher: OperationDispatcher) {
```
Comment (NachoSoto) [inline Sources/Purchasing/StoreKit1/StoreKit1Wrapper.swift:66]: I need to pass the same one down from `Purchases` here.

File: Tests/UnitTests/Purchasing/StoreKit1WrapperTests.swift [hunk 2]
```diff
@@ -253,3 +273,14 @@ class StoreKit1WrapperTests: TestCase, StoreKit1WrapperDelegate {
 #endif
 
 }
+
+private extension StoreKit1WrapperTests {
+
+    static func transaction(with: SKPayment) -> MockTransaction {
+        let transaction = MockTransaction()
+        transaction.mockPayment = SKPayment.init(product: SK1Product())
```
Comment (NachoSoto) [inline Tests/UnitTests/Purchasing/StoreKit1WrapperTests.swift:281]:
```suggestion
        transaction.mockPayment = SKPayment(product: SK1Product())
```

File: Tests/UnitTests/Purchasing/StoreKit1WrapperTests.swift [hunk 3]
```diff
@@ -87,6 +93,20 @@ class StoreKit1WrapperTests: TestCase, StoreKit1WrapperDelegate {
...
+    func testCallsDelegateToProcessTransactionsOnWorkerThread() {
+        let payment = SKPayment(product: SK1Product())
+        self.wrapper?.add(payment)
+
+        let transactions = [
+            Self.transaction(with: payment),
+            Self.transaction(with: payment)
+        ]
+
+        self.wrapper?.paymentQueue(paymentQueue, updatedTransactions: transactions)
```
Comment (NachoSoto) [inline Tests/UnitTests/Purchasing/StoreKit1WrapperTests.swift:105]:
```suggestion
        self.wrapper.paymentQueue(paymentQueue, updatedTransactions: transactions)
```

File: Tests/UnitTests/Purchasing/StoreKit1WrapperTests.swift [hunk 4]
```diff
@@ -87,6 +93,20 @@ class StoreKit1WrapperTests: TestCase, StoreKit1WrapperDelegate {
         expect(self.updatedTransactions).to(contain(transaction))
     }
 
+    func testCallsDelegateToProcessTransactionsOnWorkerThread() {
+        let payment = SKPayment(product: SK1Product())
+        self.wrapper?.add(payment)
```
Comment (NachoSoto) [inline Tests/UnitTests/Purchasing/StoreKit1WrapperTests.swift:98]:
```suggestion
        self.wrapper.add(payment)
```

**Review by aboedo — APPROVED** [review]
makes sense, good catch

**Pull request merged**